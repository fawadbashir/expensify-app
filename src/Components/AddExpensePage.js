import React, { Component } from 'react';
import Typography from '@material-ui/core/Typography'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense)
        this.props.history.push('/dashboard')
    }

    render() {
        return (
            <div>
            {/* <Typography variant="h3" color="blue">Some text in here</Typography> */}
               
                <Typography variant = 'h3'>Add Expense</Typography>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
                
            </div>
        )
    }
}



const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage)