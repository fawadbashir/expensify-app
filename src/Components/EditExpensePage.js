import React, { Component } from 'react';
import { connect } from 'react-redux'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends Component {
    onSubmit = (expense)=>{
        this.props.startEditExpense(this.props.expense.id,expense)
        this.props.history.push('/')
    }

    onClick = () => {
        this.props.startRemoveExpense({ id : this.props.expense.id})
        
        this.props.history.push('/')    
    }
    render() {
        return (
            <div>
        <ExpenseForm
        expense = {this.props.expense}
            onSubmit = {this.onSubmit}
        />
        <button onClick = {this.onClick}>Remove</button>
    </div>
        )
    }
}

// const EditExpensePage = (props) =>{
    
//     return (
//         <div>
//         <ExpenseForm
//         expense = {props.expense}
//             onSubmit = {(expense)=>{
//                 props.dispatch(editExpense(props.expense.id,expense))
//                 props.history.push('/')
//             }}
//         />
//         <button onClick = {() => {
//             props.dispatch(removeExpense({ id : props.expense.id})
//             )
//             props.history.push('/')    
//         }}
            
//             >Remove</button>
//     </div>
//     )
// }
const mapStateToProps = (state,props) => {
    return {
        expense : state.expenses.find((expense)=>expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => 
    ({startEditExpense: (expense) => dispatch(startEditExpense(expense)),
        startRemoveExpense: (expense) => dispatch(startRemoveExpense({id : expense.id}))})


export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)