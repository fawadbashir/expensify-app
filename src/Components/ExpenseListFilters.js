import React, { Component } from 'react'
import { connect } from 'react-redux'
import {DateRangePicker} from 'react-dates'
import { setTextFilter, sortByDate, sortByAmount,setStartDate,setEndDate } from '../actions/filters'
import uuidv4 from 'uuidv4'

class expenseListFilters extends Component {
    state = {
        focusedInput : null
    }
    onDatesChange =({startDate,endDate})=>{
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }
    onFocusChange = (focusedInput)=>{this.setState(()=>({focusedInput}))}

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={(e) => {
                   this.props.dispatch(setTextFilter(e.target.value))
                }} />
                <select value={this.props.filters.sortBy} onChange={(e) => {
                    if (e.target.value === 'date') {
                        this.props.dispatch(sortByDate())
                    } else {
                        this.props.dispatch(sortByAmount())
                    }
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                startDate = {this.props.filters.startDate}
                startDateId = {uuidv4()}
                endDate = {this.props.filters.endDate}
                endDateId= {uuidv4()}
                onDatesChange = {this.onDatesChange}
                focusedInput ={this.state.focusedInput}
                onFocusChange = {this.onFocusChange}
                isOutsideRange ={()=> false}

                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
export default connect(mapStateToProps)(expenseListFilters)