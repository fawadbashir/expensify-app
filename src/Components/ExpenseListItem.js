import React from 'react'
import moment from 'moment'
import numeral from 'numeral'
import { Link } from 'react-router-dom'
import Card from 'react-bootstrap/Card'



const ExpenseListItem = ({ id, description, amount, createdAt }) => (


    <Card.Body style={{ borderBottom: '1px solid #e5e5e5' }} >

        <Link className='list-group-item-action' to={`/edit/${id}`}>
            <Card.Title bg='dark' as='h3'>{description}</Card.Title>
        </Link>



        <Card.Subtitle className='mt-4'> {moment(createdAt).format('MMMM do YYYY')}
            <Card.Text as='span' className='float-right '>{numeral(amount / 100).format('$0,0.00')}</Card.Text>
        </Card.Subtitle>



    </Card.Body>





)
export default ExpenseListItem