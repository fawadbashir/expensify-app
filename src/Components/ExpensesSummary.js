import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expenses-total'

export  const ExpensesSummary = (props) => {
    const Total = numeral(expensesTotal(props.expenses) / 100).format('$0,0.00')
    const expenseCount = props.expenses.length
    const expenseWord = expenseCount === 1 ? 'Expense' : 'Expenses'
    
    return(
        <div><h4>Viewing {props.expenses.length} {expenseWord} of Total {Total}</h4>
            
        </div>
    )
} 


const mapStateToProps = (state) =>{
   return {
       expenses : selectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)