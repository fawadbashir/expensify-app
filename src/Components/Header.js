import React from 'react';
import { connect } from 'react-redux'
import {NavLink} from 'react-router-dom'
import { startLogout } from '../actions/auth'

 export const Header = (props) => {
    
     return(
        <header>
    <h1>Expensify</h1>   
     <NavLink activeClassName = 'is-active' to = '/dashboard'>Home</NavLink>      
    <NavLink activeClassName = 'is-active' to = '/create'>Add Expense </NavLink>   
    <NavLink  activeClassName = 'is-active' to = '/help'>Help</NavLink>   
    <button onClick = {props.startLogout}>LogOut</button>
    <p>{props.name}</p>
    </header>
     )
     }
    


const mapDispatchToProps = () => ({
    startLogout : startLogout()
    
})

    


export default connect(mapDispatchToProps)(Header)