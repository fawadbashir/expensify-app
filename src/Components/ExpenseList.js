import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

const ExpenseList =  (props) =>{
    return <div>
        <h1>ExpenseList</h1> 
            {/* <p>{props.expenses.length}</p> */}
            {props.expenses.map((expense)=> 
            <ExpenseListItem key = {expense.id} {...expense}/>)}
    </div>
}


const mapStatetoProps = (state) => {
    return {
        expenses : selectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStatetoProps)(ExpenseList)