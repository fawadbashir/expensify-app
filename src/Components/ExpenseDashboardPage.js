import React from 'react';
import { Link } from 'react-router-dom'
import Button from 'react-bootstrap/Button'


import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
    <div>
        <div className='pt-3 full-width-body' style={{ marginBottom: '0.5rem', padding: '0.5rem 0', }}>
            <ExpensesSummary />
            <Link to='/create'><Button size='lg' className=' mt-3' style={{ fontSize: '18px', backgroundColor: '#3D9970', color: 'white', border: 'none', padding: '10px' }} >Add Expense</Button></Link>
        </div>
        <ExpenseListFilters />
        <ExpenseList />



    </div >
)

export default ExpenseDashboardPage