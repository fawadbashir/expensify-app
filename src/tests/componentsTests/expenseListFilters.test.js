import React from 'react'
import { shallow } from 'enzyme'
import {createSerializer} from 'enzyme-to-json'
import moment from 'moment'
import { filters, altFilters} from '../fixtures/filtersFixture'
import {expenseListFilters as ExpenseListFilters} from '../../Components/ExpenseListFilters'
expect.addSnapshotSerializer(createSerializer({mode: 'shallow'}));

let setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, wrapper

beforeEach(() => {
    setTextFilter = jest.fn()
    sortByDate = jest.fn()
    sortByAmount = jest.fn()
    setStartDate = jest.fn()
    setEndDate = jest.fn()
     wrapper = shallow(
        <ExpenseListFilters
            filters = {filters}
            setTextFilter = {setTextFilter}
            sortByDate = {sortByDate}
            sortByAmount = {sortByAmount}
            setStartDate = {setStartDate}
            setEndDate = {setEndDate}
        />
    )
})

test('should render ExpenseListFilters correcty', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseListFilters with altFilters correcty', () => {
    wrapper.setProps({filters: altFilters})
    expect(wrapper).toMatchSnapshot()
})

test('should handle text Change', () => {
    const value = 'rent'
    wrapper.find('input').simulate('change', {
        target : {    value
        }
    })
    expect(setTextFilter).toHaveBeenLastCalledWith(value)
})

test('should sortBy date', () => {
    const value = 'date'
    wrapper.find('select').simulate('change', {
        target : { value}
    })
    expect(sortByDate).toHaveBeenLastCalledWith()
})

test('should sortBy amount', () => {
    const value = 'amount'
    wrapper.find('select').simulate('change', {
        target : { value }
    })
    expect(sortByAmount).toHaveBeenCalled()
})

test('should handle date Changes', () => {
    const startDate = moment(0).add(4, 'years')
    const endDate = moment(0).add(8, 'years')
    wrapper.find('withStyles(DateRangePicker)').prop('onDatesChange')({startDate, endDate})
    expect(setStartDate).toHaveBeenLastCalledWith(startDate)
    expect(setEndDate).toHaveBeenLastCalledWith(endDate)
})

test('should handle Focus Change', () => {
    wrapper.find('withStyles(DateRangePicker)').prop('onFocusChange')('endDate')
    expect(wrapper.state('focusedInput')).toBe('endDate')
})