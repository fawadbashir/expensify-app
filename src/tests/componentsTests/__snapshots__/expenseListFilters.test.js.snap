// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters correcty 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={null}
    endDateId="e10790f9-49a4-4cab-a656-3c7a62d5ad14"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={2}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={false}
    showDefaultInputIcon={false}
    small={false}
    startDate={null}
    startDateId="4aa25734-be9c-427f-9c97-6192cba78b8c"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;

exports[`should render ExpenseListFilters with altFilters correcty 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value="bills"
  />
  <select
    onChange={[Function]}
    value="amount"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={"1970-01-04T00:00:00.000Z"}
    endDateId="1a9a5c1e-25d5-4392-af53-42d30da9276a"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={2}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={false}
    showDefaultInputIcon={false}
    small={false}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="4396b1b0-7aaa-4a89-a25a-9c239840bb3b"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
